odmantic/engine.py

```diff
@@ -511,28 +551,37 @@ async def find_one(
             return None
         return results[0]
 
+    async def _save_collection_updates(
+        self,
+        collection_updates: CollectionUpdatesType,
+        session: "AsyncIOMotorClientSession",
+    ) -> None:
+        # reverse so that the last collections added, the ones for sub-documents, are
+        # saved first
+        for collection_name, updates in reversed(list(collection_updates.items())):
+            collection = self._get_collection_from_name(collection_name)
+            update_operations = [update[0] for update in updates]
+            update_instances = [update[1] for update in updates]
+            await collection.bulk_write(
+                update_operations,
+                session=session,
+            )
+            for inst in update_instances:
+                object.__setattr__(inst, "__fields_modified__", set())
+
     async def _save(
         self, instance: ModelType, session: "AsyncIOMotorClientSession"
     ) -> ModelType:
         """Perform an atomic save operation in the specified session"""
-        for ref_field_name in instance.__references__:
-            sub_instance = cast(Model, getattr(instance, ref_field_name))
-            await self._save(sub_instance, session)
-
-        fields_to_update = instance.__fields_modified__ | instance.__mutable_fields__
-        if len(fields_to_update) > 0:
-            doc = instance.doc(include=fields_to_update)
-            collection = self.get_collection(type(instance))
-            try:
-                await collection.update_one(
-                    instance.doc(include={instance.__primary_field__}),
-                    {"$set": doc},
-                    upsert=True,
-                    session=session,
-                )
-            except pymongo.errors.DuplicateKeyError as e:
+        collection_updates = self._prepare_document_updates(instance)
+        try:
+            await self._save_collection_updates(
+                collection_updates=collection_updates, session=session
+            )
+        except pymongo.errors.BulkWriteError as e:
+            if e.details["writeErrors"][0]["code"] == 11000:
                 raise DuplicateKeyError(instance, e)
-            object.__setattr__(instance, "__fields_modified__", set())
+            raise  # pragma: no cover
```

Comment:
Maybe we could try to reach this, WDYT ?

Reply:
I have been trying to trigger an error that shows up here for the past hours but I haven't figured out something that would error out here and that we can trigger in tests without monkeypatching everything (that would defeat the purpose of the test).

But maybe there's a simple thing we could test that would certainly raise here and I just don't know it yet.

If you have any ideas, let me know!

Reply:
I have given up about reaching this `raise` 😔 

I've been two days trying to find something that would trigger a `BulkWriteError` that is not a `DuplicateKeyError` and that we can trigger from some simple update in ODMantic models. I haven't found anything, there might not be any other type of error, but there's no way to be sure, and if we don't raise here, we would end up swallowing any other potential or future exception there.

But I can't find something that would raise something that goes to this code path, I tried a lot of things and got a lot of different exceptions, but nothing inside of this code path.

I even tried looking a bit at the source code of MongoDB itself, looking for those keys, but there's nothing obvious that gives away what other types of errors could end up there.

If you have any ideas, please let me know! (or feel free to take over and add commits).

Reply:
Long time :upside_down_face: . It seems hard to reach this line.